src/provider.py

```diff
@@ -0,0 +1,185 @@
+
+        return relation_config
+
+    def update_acls(self, event: RelationChangedEvent) -> None:
+        """Updates cluster ACLs for a given event relation to match client relation data.
+
+        Args:
+            event: the event from a related client application needing ACLs
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        if not self.charm.kafka_config.zookeeper_connected:
+            logger.debug("cannot update ACLs, ZooKeeper not yet connected")
+            event.defer()
+            return
+
+        provider_relation_config = self.provider_relation_config(event=event)
+        requirer_relation_config = self.requirer_relation_config(event=event)
+
+        self.kafka_auth.load_current_acls()
+
+        self.kafka_auth.update_user_acls(
+            username=provider_relation_config["username"],
+            group=provider_relation_config.get("consumer-group-prefix"),
+            **requirer_relation_config,
+        )
+
+        # non-leader units need cluster_config_changed event to update their super.users
+        self.charm.model.get_relation(PEER).data[self.charm.app].update(
+            {"super-users": self.kafka_config.super_users}
+        )
+
+        event.relation.data[self.charm.app].update(provider_relation_config)
+
+    def _on_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handler for `kafka-client-relation-created` event.
+
+        Adds new relation users to ZooKeeper.
+
+        Args:
+            event: the event from a related client application needing a user
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        if not self.charm.ready_to_start:
+            logger.debug("cannot add user, ZooKeeper not yet connected")
+            event.defer()
+            return
+
+        provider_relation_config = self.provider_relation_config(event=event)
+
+        self.kafka_auth.add_user(
+            username=provider_relation_config["username"],
+            password=provider_relation_config["password"],
+        )
+
+        # non-leader units need cluster_config_changed event to update their super.users
+        self.charm.model.get_relation(PEER).data[self.charm.app].update(
+            {provider_relation_config["username"]: provider_relation_config["password"]}
+        )
+
+    def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handler for `kafka-client-relation-broken` event.
+
+        Removes relation users from ZooKeeper.
+
+        Args:
+            event: the event from a related client application needing a user
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        if not self.charm.ready_to_start:
+            logger.debug("cannot remove user, ZooKeeper not yet connected")
+            event.defer()
+            return
+
+        if (
+            event.relation.app != self.charm.app or not self.charm.app.planned_units() == 0
```

Comment:
I found that sometimes the planned units is not set to zero when removing the application. Should that be a problem in this check?

Reply:
What causes that? In this case, I have the following related apps -- `ZooKeeper <-> Kafka <-> Client`

When `juju remove-application kafka`, Kafka gets 2 relation broken events, one for Client and one for ZooKeeper.

I want to handle `client_relation_broken` (i.e to remove users/ACLs), but only when it's not Kafka that's being removed. In that case, I want to ignore it. 

Do you know of a good way of handling that?

Reply:
I haven't figured out why and if there is a specific situation it happens.

Thanks for the details, Marc. I don't have a good way of handling this situation. I had a similar situation, but in that I had to avoid deleting the database user if the postgres app is being scaled down. It was a little bit different and the solution was just a workaround, not a recommendation (I used some relation data between relation departed and relation broken events). My workaround is available at https://github.com/canonical/postgresql-operator/blob/ceeba024995a33714071fbf59e9fe759d6394d71/src/relations/postgresql_provider.py#L109 and https://github.com/canonical/postgresql-operator/blob/ceeba024995a33714071fbf59e9fe759d6394d71/src/relations/postgresql_provider.py#L128, but from what I remember maybe it wouldn't work when removing an application.